### PR TITLE
added support for sampling

### PR DESCRIPTION
**Pull request by hemanthk269**
When specifying sampling for a dataset, it can be provided in two ways:
1. **Simply specify the sampling rate** when you want to sample the dataset without specifying the columns used for sampling.
   - **Sampling Rate**: A float between 0 and 1 that determines the proportion of the dataset to include.
2. **Use the `Sample` object** when you want to specify both the sampling rate and the specific columns used for sampling.
   - **Sampling Rate**: A float between 0 and 1 that determines the proportion of the dataset to include.
   - **Using**: A list of columns used to hash for sampling the data. Preprocessing columns and the timestamp field cannot be included in this list.

Default Behavior When No Columns Are Specified
1. For Keyed Datasets:
All key columns are used for sampling, excluding any preproc columns.
2. For Non-Keyed Datasets:
All columns are used for sampling except for the timestamp and preproc columns.

Added unit tests.
Updated documentation.

Screenshot 2024-10-22 at 2 22 13 PM

----

> [!IMPORTANT]
> Adds support for dataset sampling with configurable rate and columns, updating connectors, protobufs, tests, and documentation.
> 
>   - **Behavior**:
>     - Adds `Sample` class in `connectors.py` to define sampling rate and fields for hashing.
>     - Updates `source()` function in `connectors.py` to accept `sample` parameter, validating its type and range.
>     - Integrates sampling into source connectors in `to_proto.py`, including `_sample_to_proto()` function.
>   - **Protobuf**:
>     - Adds `SamplingStrategy` message in `connector_pb2.py` and `connector_pb2.pyi`.
>     - Updates `Source` message in `connector_pb2.py` to include `sampling_strategy`.
>   - **Tests**:
>     - Adds `test_source_with_sample()` in `test_connectors.py` to test sampling functionality.
>   - **Documentation**:
>     - Updates `source.md` to document the `sample` parameter.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=fennel-ai%2Fclient&utm_source=github&utm_medium=referral)<sup> for 71143d746ca8aa4d52c67e1c48746545beb9ac70. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: fennel/connectors/connectors.py [hunk 1]
```diff
@@ -106,6 +117,7 @@ def source(
     bounded: bool = False,
     idleness: Optional[Duration] = None,
     where: Optional[Callable] = None,
+    sample: Optional[Union[float, Sample]] = None,
```
Comment (saiharshavellanki) [inline fennel/connectors/connectors.py:120]: Why not Optional[Sample]?

Comment (hemanthk269) [reply fennel/connectors/connectors.py:120]: User can give float directly without specifying any columns
https://coda.io/d/Sampled-Ingestion_dkplIi5d_Kp/Sampled-Ingestion_suIbBFju#_lu9Aph4j

File: fennel/connectors/connectors.py [hunk 2]
```diff
@@ -184,6 +198,18 @@ def source(
             "idleness parameter should not be passed when bounded is set as False"
         )
 
+    if sample is not None:
```
Comment (saiharshavellanki) [inline fennel/connectors/connectors.py:201]: Can we also verify that preproc columns shouldn't be part of sampling columns here?

File: fennel/internal_lib/to_proto/to_proto.py [hunk 3]
```diff
@@ -1104,6 +1113,78 @@ def _renames_to_proto(renames: Optional[Dict[str, str]]) -> Mapping[str, str]:
     return renames
 
 
+def get_sampling_rate(sample: Optional[Union[float, Sample]]):
+    if isinstance(sample, float):
+        if sample < 0 or sample > 1:
```
Comment (saiharshavellanki) [inline fennel/internal_lib/to_proto/to_proto.py:1118]: Why to add validation again here? Is the lint failing?

Comment (hemanthk269) [reply fennel/internal_lib/to_proto/to_proto.py:1118]: yes

File: fennel/internal_lib/to_proto/to_proto.py [hunk 4]
```diff
@@ -1104,6 +1113,78 @@ def _renames_to_proto(renames: Optional[Dict[str, str]]) -> Mapping[str, str]:
...
+        return sample.rate
+    raise ValueError(f"Invalid sample type: {type(sample)}")
+
+
+def get_sampling_columns(
+    sample: Optional[Union[float, Sample]],
+    dataset: Dataset,
+    preproc: Optional[Dict[str, PreProcValue]],
+):
+    disallowed_columns = [dataset.timestamp_field]
```
Comment (saiharshavellanki) [inline fennel/internal_lib/to_proto/to_proto.py:1133]: lets make this empty given we have separate check for timestamp column below

Comment (hemanthk269) [reply fennel/internal_lib/to_proto/to_proto.py:1133]: we are using this when user sent only sampling_rate as well.

File: fennel/internal_lib/to_proto/to_proto.py [hunk 5]
```diff
@@ -1104,6 +1113,78 @@ def _renames_to_proto(renames: Optional[Dict[str, str]]) -> Mapping[str, str]:
...
+                )
+        return sorted(input_columns)
+
+    if len(dataset.key_fields) != 0:
+        key_columns = dataset.key_fields
+        return sorted(
+            [
+                column
+                for column in key_columns
+                if column not in disallowed_columns
```
Comment (saiharshavellanki) [inline fennel/internal_lib/to_proto/to_proto.py:1161]: We should raise an error here given we are skipping some key fields

File: fennel/internal_lib/to_proto/to_proto.py [hunk 6]
```diff
@@ -1104,6 +1113,78 @@ def _renames_to_proto(renames: Optional[Dict[str, str]]) -> Mapping[str, str]:
...
+            ]
+        )
+    else:
+        all_columns = [str(column) for column in dataset.fields]
+        return sorted(
+            [
+                column
+                for column in all_columns
+                if column not in disallowed_columns
```
Comment (saiharshavellanki) [inline fennel/internal_lib/to_proto/to_proto.py:1170]: We should raise an error here given we are skipping some key fields to ask customer to add sampling columns

File: fennel/connectors/connectors.py (hunk 2 above)
Comment (hemanthk269) [reply fennel/connectors/connectors.py:201]: we are doing that in to_proto file, is there any reason to do it in 2 places?

File: fennel/internal_lib/to_proto/to_proto.py (hunk 5 above)
Comment (hemanthk269) [reply fennel/internal_lib/to_proto/to_proto.py:1161]: if customer just want to sample the dataset, is it ok to give exception here?

**Review by saiharshavellanki — APPROVED** [review]

File: fennel/connectors/connectors.py (hunk 2 above)
Comment (hemanthk269) [reply fennel/connectors/connectors.py:201]: Added that check

File: fennel/internal_lib/to_proto/to_proto.py (hunk 5 above)
Comment (hemanthk269) [reply fennel/internal_lib/to_proto/to_proto.py:1161]: throwing exception when all key columns are derived from preproc.

File: fennel/internal_lib/to_proto/to_proto.py (hunk 6 above)
Comment (hemanthk269) [reply fennel/internal_lib/to_proto/to_proto.py:1170]: throwing exception if all columns except timestamp are part of preproc.

**Pull request merged**